crates/subspace-core-primitives/src/lib.rs

```diff
@@ -95,23 +93,122 @@ pub type Blake3Hash = [u8; BLAKE3_HASH_SIZE];
     Hash,
     From,
     Into,
+    AsRef,
+    AsMut,
     Deref,
     DerefMut,
     Encode,
     Decode,
     TypeInfo,
     MaxEncodedLen,
 )]
+#[into(ref, ref_mut)]
+#[as_ref(forward)]
+#[as_mut(forward)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 #[cfg_attr(feature = "serde", serde(transparent))]
-pub struct Blake3HashHex(#[cfg_attr(feature = "serde", serde(with = "hex"))] Blake3Hash);
+pub struct Blake3Hash(#[cfg_attr(feature = "serde", serde(with = "hex"))] [u8; BLAKE3_HASH_SIZE]);
 
-impl fmt::Debug for Blake3HashHex {
+impl fmt::Debug for Blake3Hash {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         write!(f, "{}", hex::encode(self.0))
     }
 }
 
+impl FromHex for Blake3Hash {
+    type Error = hex::FromHexError;
+
+    fn from_hex<T: AsRef<[u8]>>(hex: T) -> Result<Self, Self::Error> {
+        let data = hex::decode(hex)?
+            .try_into()
+            .map_err(|_| hex::FromHexError::InvalidStringLength)?;
+
+        Ok(Self(data))
+    }
+}
+
+impl From<Blake3Hash> for Vec<u8> {
+    #[inline]
+    fn from(hash: Blake3Hash) -> Self {
+        hash.0.to_vec()
+    }
+}
+
+impl From<Blake3Hash> for PosSeed {
+    #[inline]
+    fn from(hash: Blake3Hash) -> Self {
+        Self(hash.0)
+    }
+}
+
+impl From<Blake3Hash> for Randomness {
+    #[inline]
+    fn from(hash: Blake3Hash) -> Self {
+        Self::from(hash.0)
+    }
+}
+
+impl From<&[u8; BLAKE3_HASH_SIZE]> for Blake3Hash {
+    #[inline]
+    fn from(value: &[u8; BLAKE3_HASH_SIZE]) -> Self {
+        Self(*value)
+    }
+}
+
+impl TryFrom<&[u8]> for Blake3Hash {
+    type Error = TryFromSliceError;
+
+    #[inline]
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        Ok(Self(value.try_into()?))
+    }
+}
+
+impl TryFrom<Vec<u8>> for Blake3Hash {
+    type Error = TryFromSliceError;
+
+    #[inline]
+    fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
+        value.as_slice().try_into()
+    }
+}
+
+impl PartialEq<&[u8]> for Blake3Hash {
+    #[inline]
+    fn eq(&self, other: &&[u8]) -> bool {
+        self.0 == *other
+    }
+}
+
+impl PartialEq<Blake3Hash> for &[u8] {
+    #[inline]
+    fn eq(&self, other: &Blake3Hash) -> bool {
+        *self == other.0
+    }
+}
+
+impl Blake3Hash {
+    /// Creates a new hash from an array.
+    /// Used to resolve type ambiguity.
+    #[inline]
+    pub fn from_array(array: [u8; BLAKE3_HASH_SIZE]) -> Self {
+        Self(array)
+    }
+
+    /// Creates a new hash from an array reference.
+    /// Used to resolve type ambiguity.
+    #[inline]
+    pub fn from_array_ref(array_ref: &[u8; BLAKE3_HASH_SIZE]) -> Self {
+        Self(*array_ref)
+    }
+
+    /// Returns this hash as an owned array.
+    #[inline]
+    pub fn as_array(self) -> [u8; BLAKE3_HASH_SIZE] {
+        self.0
+    }
```

Comment:
Similarly this is just `.as_ref()` that we already have or even `&` due to dereferencing:
```suggestion
```

Reply:
Yes, but only when the types are unambiguous. I'll try it and you can see what you think.

Reply:
If you need an array specifically then you can do de-referencing `*value` or maybe `**value` I didn't test which one compiles. Either way this is not really needed as a standalone method, we have plenty of types that encapsulate `[u8; 32]` or `[u8; 48]` without such methods.